crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+                            "found at least one matching server cert" :
+                            "no current matching server cert found");
+    return 1;
+}
+
+/*
+ * Exceptional handling for 3GPP TS 33.310, only to use for IP and if the ctx
+ * option is explicitly set: use self-signed certificates from extraCerts as
+ * trust anchor to validate server cert - provided it also can validate the
+ * newly enrolled certificate
+ */
+static int srv_cert_valid_3gpp(OSSL_CMP_CTX *ctx, X509 *scrt,
+                               const OSSL_CMP_MSG *msg)
+{
+    int valid = 0;
+    X509_STORE *store = X509_STORE_new();
+
+    if (store != NULL /* store does not include CRLs */
+            && ossl_cmp_X509_STORE_add1_certs(store, msg->extraCerts,
+                                              1 /* self-signed only */))
+        valid = OSSL_CMP_validate_cert_path(ctx, store, scrt);
+
+    if (valid) {
+        /*
+         * verify that the newly enrolled certificate (which is assumed to have
+         * rid == 0) can also be validated with the same trusted store
+         */
+        EVP_PKEY *privkey = OSSL_CMP_CTX_get0_newPkey(ctx, 1);
+        OSSL_CMP_CERTRESPONSE *crep =
+            ossl_cmp_certrepmessage_get0_certresponse(msg->body->value.ip, 0);
+        X509 *newcrt = ossl_cmp_certresponse_get1_certificate(privkey, crep);
+        /*
+         * maybe better use get_cert_status() from cmp_client.c, which catches
+         * errors
+         */
+        valid = OSSL_CMP_validate_cert_path(ctx, store, newcrt);
+        X509_free(newcrt);
+    }
+    X509_STORE_free(store);
+    return valid;
+}
+
+/* Find and validate any potentially usable server cert */
+static X509 *find_srvcert(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+{
+    X509 *scrt = NULL;
+    int valid = 0;
+    GENERAL_NAME *sender = msg->header->sender;
+
+    if (sender == NULL || msg->body == NULL)
+        return 0; /* other NULL cases already have been checked */
+    if (sender->type != GEN_DIRNAME) {
+        CMPerr(0, CMP_R_SENDER_GENERALNAME_TYPE_NOT_SUPPORTED);
+        return NULL; /* FR#42: support for more than X509_NAME */
+    }
+
+    /*
+     * valid scrt, matching sender name, found earlier in transaction, will be
+     * used for validating any further msgs where extraCerts may be left out
+     */
+    (void)ERR_set_mark();
+    if (ctx->validatedSrvCert != NULL
+            && cert_acceptable(ctx->validatedSrvCert, msg, ctx->trusted)) {
+        (void)ERR_pop_to_mark();
+        scrt = ctx->validatedSrvCert;
+        valid = 1;
+    } else {
+        STACK_OF(X509) *found_crts;
+        char *sname;
+        int i;
+
+        (void)ERR_pop_to_mark();
+
+        /* sk_TYPE_find to use compfunc X509_cmp, not ptr comparison */
+        if ((found_crts = sk_X509_new_null()) == NULL)
+            return NULL;
+
+        /* tentatively set error, which allows accumulating diagnostic info */
+        (void)ERR_set_mark();
```

Comment:
We set a mark here, but not all codepaths call ERR_pop_to_mark(). Probably we should either call that or ERR_clear_last_mark() in all cases.

Reply:
I added the call to ERR_clear_last_mark().